jupyterlab_gcedetails/src/service/price_service.ts

```diff
@@ -0,0 +1,219 @@
+    text: 'asia-southeast1 (Singapore)',
+  },
+];
+
+export class PriceService {
+  private catalog: Catalog;
+  private _receivedError: boolean;
+
+  async getPriceList(): Promise<void> {
+    // Only rebuild ctalog if it hasn't been built or was built unsuccessfully
+    if (!this.catalog || this._receivedError) {
+      try {
+        this.catalog = new Map();
+        const response = await ServerConnection.makeRequest(
+          `${ServerConnection.defaultSettings.baseUrl}gcp/v1/pricelist`,
+          {},
+          ServerConnection.defaultSettings
+        );
+        if (!response.ok) {
+          throw { result: await response.text() };
+        }
+        const fullCatalog = await response.json();
+        Object.keys(fullCatalog).forEach(sku => {
+          if (sku.endsWith('PREEMPTIBLE')) return;
+          if (
+            sku.startsWith('CP-COMPUTEENGINE-VMIMAGE-N1') ||
+            sku.startsWith('CP-COMPUTEENGINE-VMIMAGE-E2')
+          ) {
+            this.populateCatalog(
+              sku.substring(25).toLowerCase(),
+              fullCatalog[sku],
+              true
+            );
+          } else if (sku.startsWith('GPU_')) {
+            this.populateCatalog(sku.substring(4), fullCatalog[sku], true);
+          } else if (sku === 'CP-COMPUTEENGINE-STORAGE-PD-CAPACITY') {
+            this.populateCatalog('pd-standard', fullCatalog[sku], false);
+          } else if (sku === 'CP-COMPUTEENGINE-STORAGE-PD-SSD') {
+            this.populateCatalog('pd-ssd', fullCatalog[sku], false);
+          }
+        });
+        this._receivedError = false;
+      } catch (err) {
+        console.error(`Unable to retrieve pricelist.`);
+        this._receivedError = true;
+      }
+    }
+  }
+
+  get receivedError() {
+    return this._receivedError;
+  }
+
+  // tslint:disable-next-line:no-any The json object returned is not typed.
+  private populateCatalog(sku: string, prices: any, timeBased: boolean) {
+    Object.keys(prices).forEach(region => {
+      const regionCatalog = this.getRegionCatalog(region);
+      if (regionCatalog) {
+        regionCatalog.set(
+          sku,
+          prices[region] * (timeBased ? HOURS_PER_MONTH : 1.0)
+        );
+      }
+    });
+  }
+
+  private getRegionCatalog(region: string): Map<string, number> | undefined {
+    if (REGIONS.find(r => r.value === region)) {
+      if (!this.catalog.has(region)) {
+        this.catalog.set(region, new Map<string, number>());
+      }
+      return this.catalog.get(region);
+    }
+    return undefined;
+  }
+
+  private getRegionFromZone(zone: string): string {
+    return zone.split('-', 2).join('-');
+  }
+
+  /**
+   * Returns estimated total monthly price for a given instance to be created in a
+   * certain region.
+   *
+   * Logics adapted from
+   * //depot/google3/java/com/google/developers/console/web/common/compute/pricing/pricing_service.ts
+   */
+  getPrice(zone: string, config: HardwareConfiguration): number {
```

Comment:
The signature for this should be number|undefined.

So the more I think about this, the more I think that this should be the only public method on this service. In Angular, we have the concept of route resolvers that load all of the data needed for a particular page. Here, we are probably OK with waiting until a user explicitly wants to make a hardware config change before fetching the data.

I'm thinking that the actual price retrieval logic is private, and that when this method is called, we use a private Promise data member that has the Catalog. If the Catalog is empty, this method will return undefined as it currently does, otherwise the same logic will apply.

In this way, we won't need to preload the data or check for an error state. It'll be a best attempt to determine pricing, but if we can't load the data for some reason or the configuration doesn't exist, we'll show nothing.